### DISCUSSION TITLE
Rewrite implementation of FastF1 API?

### DISCUSSION THREAD

**Discussion by pbulsink** [SCasanova/f1dataR#206] [Ideas]
I'm wondering if in the off-season we expand/rewrite some of the capabilities of f1dataR to match the new capabilities of FastF1, and also be more capable of implementing changes to reflect the changes there (#198).

For example, right now, for us to add and fully use `pick_drivers` instead of using `pick_driver` takes a rewrite of our `get_session_laps()` (#78). Had `pick_driver()` instead been a standalone function (as part of a class or otherwise) it'd be much easier to just accept a vector of character strings or numbers instead of a single value.

I also wonder if having an interface that more closely matches the workflow of FastF1 will make it easier for people to duplicate or expand on analysis they otherwise see happening in Python?

Thoughts?

**Comment by SCasanova** [comment]
I'm al for it. To be honest the original implementations didn't have much though put into it and I didn't know what the package would turn into. Big fan of rewriting old code

**Comment by pbulsink** [comment]
Played around with fastf1 for a bit with reticulate. 

If you start with:
```r
reticulate::py_run_string("import fastf1")
ff1s <- reticulate::py_run_string("session = fastf1.get_session(2023, 'Brazil', 'R')")
```

Then you can call the rest of the normal fastf1 python code on that session object.
For example:
`ff1s$session$load()`
then
- `ff1s$session$laps` will return a data.frame of laps
- `ff1s$session$drivers` will return a vector of driver numbers in the session
- `ff1s$session$get_circuit_info()` will pull the [newly implemented additional circuit info](https://docs.fastf1.dev/core.html#fastf1.core.Session.get_circuit_info)

If you want to use lap pick functions it gets a bit more confusing:
`ff1s$fastf1$core$Laps$pick_drivers(self = ff1s$session$laps, identifiers = c("VER", "LEC"))` 
returns the two drivers' lap times. If you've previously saved `ff1s$session$laps` as a R object, you can provide that for `self`, e.g. `laps <- ff1s$session$laps`, then `ff1s$fastf1$core$Laps$pick_drivers(self = laps, identifiers = c("VER", "LEC"))` works too.

I think there's likely a midpoint between wrapping all of the fastf1 functions and leaving it all to the user - I just haven't figured out that convenient midpoint yet.
